### PR TITLE
Make assertion predicates optional and repeated arguments

### PR DESCRIPTION
**Pull request by wi101**
Solves #32

### CODE REVIEW

File: src/core/probably.scala [hunk 1]
```diff
@@ -112,23 +112,31 @@ class Runner(specifiedTests: Set[TestId] = Set()) extends Dynamic {
...
     def id: TestId = TestId(Runner.shortDigest(name))
     def action(): Type
-    
-    def assert(predicate: Type => Boolean): Unit =
-      try if(runTest(id)) check(predicate) catch { case NonFatal(e) => () }
-    
-    def check(predicate: Type => Boolean): Type = {
+
+    def assert(predicate: (Type => Boolean)*): Unit =
+      try if(runTest(id)) check(predicate : _*) catch { case NonFatal(_) => () }
```
Comment (propensive) [inline src/core/probably.scala:120]:
```suggestion
      try if(runTest(id)) check(predicate: _*) catch { case NonFatal(_) => () }
```

File: src/core/probably.scala [hunk 2]
```diff
@@ -112,23 +112,31 @@ class Runner(specifiedTests: Set[TestId] = Set()) extends Dynamic {
...
+
+    def check(predicates: (Type => Boolean)*): Type = {
+      def handler(index: Int): PartialFunction[Throwable, Datapoint] = { case e: Exception => ThrowsInCheck(e, map, index) }
+      def makeDatapoint(predicates: Seq[Type => Boolean], count: Int, datapoint: Datapoint, value: Type): Datapoint =
```
Comment (propensive) [inline src/core/probably.scala:124]:
```suggestion

      def makeDatapoint(predicates: Seq[Type => Boolean], count: Int, datapoint: Datapoint, value: Type): Datapoint =
```

File: src/core/probably.scala [hunk 3]
```diff
@@ -112,23 +112,31 @@ class Runner(specifiedTests: Set[TestId] = Set()) extends Dynamic {
...
+        try {
+          if (predicates.isEmpty) datapoint
+          else
+            if (predicates.head(value)) makeDatapoint(predicates.tail, count + 1, Pass, value)
+            else Fail(map, count)
```
Comment (propensive) [inline src/core/probably.scala:129]:
```suggestion
          else if (predicates.head(value)) makeDatapoint(predicates.tail, count + 1, Pass, value)
          else Fail(map, count)
```

File: src/core/probably.scala [hunk 4]
```diff
@@ -112,23 +112,31 @@ class Runner(specifiedTests: Set[TestId] = Set()) extends Dynamic {
...
 
       value match {
         case Success(value) =>
-          def handler: PartialFunction[Throwable, Datapoint] = { case e: Exception => ThrowsInCheck(e, map) }
-          val datapoint: Datapoint = try(if(predicate(value)) Pass else Fail(map)) catch handler
-          record(this, time, datapoint)
+          record(this, time, makeDatapoint(predicates, 0, Pass, value))
           value
         case Failure(exception) =>
           record(this, time, Throws(exception, map))
```
Comment (propensive) [inline src/core/probably.scala:142]:
```suggestion
        case Failure(exception) => record(this, time, Throws(exception, map))
```

File: src/core/probably.scala [hunk 5]
```diff
@@ -112,23 +112,31 @@ class Runner(specifiedTests: Set[TestId] = Set()) extends Dynamic {
...
       val t0 = System.currentTimeMillis()
       val value = Try(action())
       val time = System.currentTimeMillis() - t0
 
       value match {
         case Success(value) =>
-          def handler: PartialFunction[Throwable, Datapoint] = { case e: Exception => ThrowsInCheck(e, map) }
-          val datapoint: Datapoint = try(if(predicate(value)) Pass else Fail(map)) catch handler
-          record(this, time, datapoint)
+          record(this, time, makeDatapoint(predicates, 0, Pass, value))
```
Comment (propensive) [inline src/core/probably.scala:139]:
```suggestion
```

File: src/core/probably.scala [hunk 6]
```diff
@@ -112,23 +112,31 @@ class Runner(specifiedTests: Set[TestId] = Set()) extends Dynamic {
...
+            else Fail(map, count)
+        } catch handler(count)
+
+
       val t0 = System.currentTimeMillis()
       val value = Try(action())
       val time = System.currentTimeMillis() - t0
 
       value match {
         case Success(value) =>
```
Comment (propensive) [inline src/core/probably.scala:138]:
```suggestion
        case Success(value) => record(this, time, makeDatapoint(predicates, 0, Pass, value))
```

File: src/test/tests.scala [hunk 7]
```diff
@@ -86,7 +88,7 @@ object Main extends Suite("Probably Tests") {
       for(i <- 1 to 10) runner("integers are less than six", i = i)(i).assert(_ < 6)
       runner.report().results.head.outcome
     }.assert {
-      case Runner.FailsAt(Runner.Fail(map), _) if map("i") == "6" => true
+      case Runner.FailsAt(Runner.Fail(map, _), _) if map("i") == "6" => true
```
Comment (propensive) [inline src/test/tests.scala:91]:
```suggestion
      case FailsAt(Fail(map, _), _) if map("i") == "6" => true
```

File: src/test/tests.scala [hunk 8]
```diff
@@ -109,5 +111,62 @@ object Main extends Suite("Probably Tests") {
...
       runner.report
     }.assert(_.results.head.count == 100)
+
+    test("assert without predicate should succeed") {
+      val runner = new Runner()
+      runner("test division") {
+        5 / 2
+      }.assert()
+      runner.report()
+    }.assert(r => r.passed == r.total)
```
Comment (propensive) [inline src/test/tests.scala:121]: Probably best to check that the total and the number of passed tests are both one.
```suggestion
    }.assert(_.passed == 1, _.total == 1)
```

File: src/test/tests.scala [hunk 9]
```diff
@@ -109,5 +111,62 @@ object Main extends Suite("Probably Tests") {
...
+
+    test("assert without predicate should fail") {
+      val runner = new Runner()
+      runner("test division") {
+        5 / 0
+      }.assert()
+      runner.report()
+    }.assert(r => r.passed == 0)
```
Comment (propensive) [inline src/test/tests.scala:129]:
```suggestion
    }.assert(_.passed == 0)
```

File: src/test/tests.scala [hunk 10]
```diff
@@ -109,5 +111,62 @@ object Main extends Suite("Probably Tests") {
...
+
+    test("check without predicate should succeed") {
+      val runner = new Runner()
+      runner("test division") {
+        5 / 2
+      }.check()
+      runner.report()
+    }.assert(r => r.passed == r.total)
```
Comment (propensive) [inline src/test/tests.scala:137]:
```suggestion
    }.assert(_.passed == 1, _.total == 1)
```

File: src/test/tests.scala [hunk 11]
```diff
@@ -109,5 +111,62 @@ object Main extends Suite("Probably Tests") {
...
+
+    test("check without predicate should fail") {
+      val runner = new Runner()
+      Try(runner("test division") {
+        5 / 0
+      }.check())
+      runner.report()
+    }.assert(r => r.failed == 1)
```
Comment (propensive) [inline src/test/tests.scala:145]:
```suggestion
    }.assert(_.failed == 1, _.total == 1)
```

File: src/test/tests.scala [hunk 12]
```diff
@@ -109,5 +111,62 @@ object Main extends Suite("Probably Tests") {
...
+
+    test("assert with 2 successful predicates") {
+      val runner = new Runner()
+      runner("test double") {
+        0.001
+      }.assert(_ >= 0.0, _ <= 1.0)
+      runner.report()
+    }.assert(r => r.passed == r.total)
```
Comment (propensive) [inline src/test/tests.scala:153]:
```suggestion
    }.assert(_.passed == 1, _.total == 1)
```

File: src/test/tests.scala [hunk 13]
```diff
@@ -109,5 +111,62 @@ object Main extends Suite("Probably Tests") {
...
+
+    test("assert with 2 predicates") {
+      val runner = new Runner()
+      runner("test double") {
+        0.001
+      }.assert(_ >= 0.0, _ < 0.0)
+      runner.report().results.head.outcome
+    }.assert(_ == FailsAt(Fail(Map.empty, 1), 1))
```
Comment (propensive) [inline src/test/tests.scala:161]:
```suggestion
    }.assert(_ == FailsAt(Fail(Map(), 1), 1))
```

File: src/test/tests.scala [hunk 14]
```diff
@@ -109,5 +111,62 @@ object Main extends Suite("Probably Tests") {
...
+
+    test("assert with 3 predicates") {
+      val runner = new Runner()
+      runner("test double") {
+        0.001
+      }.assert(_ >= 0.0, _ <= 1.0, _ < 0.0)
+      runner.report().results.head.outcome
+    }.assert(_ == FailsAt(Fail(Map.empty, 2), 1))
```
Comment (propensive) [inline src/test/tests.scala:169]:
```suggestion
    }.assert(_ == FailsAt(Fail(Map(), 2), 1))
```

File: src/test/tests.scala [hunk 15]
```diff
@@ -109,5 +111,62 @@ object Main extends Suite("Probably Tests") {
...
+        0.001
+      }.assert(_ >= 0.0, _ < 0.0)
+      runner.report().results.head.outcome
+    }.assert(_ == FailsAt(Fail(Map.empty, 1), 1))
+
+    test("assert with 3 predicates") {
+      val runner = new Runner()
+      runner("test double") {
+        0.001
+      }.assert(_ >= 0.0, _ <= 1.0, _ < 0.0)
```
Comment (propensive) [inline src/test/tests.scala:167]:
```suggestion
      runner("test double")(0.001).assert(_ >= 0.0, _ <= 1.0, _ < 0.0)
```

File: src/test/tests.scala [hunk 16]
```diff
@@ -109,5 +111,62 @@ object Main extends Suite("Probably Tests") {
...
+        0.001
+      }.assert(_ >= 0.0, _ <= 1.0)
+      runner.report()
+    }.assert(r => r.passed == r.total)
+
+    test("assert with 2 predicates") {
+      val runner = new Runner()
+      runner("test double") {
+        0.001
+      }.assert(_ >= 0.0, _ < 0.0)
```
Comment (propensive) [inline src/test/tests.scala:159]:
```suggestion
      runner("test double")(0.0010.assert(_ >= 0.0, _ < 0.0)
```

File: src/test/tests.scala [hunk 17]
```diff
@@ -109,5 +111,62 @@ object Main extends Suite("Probably Tests") {
...
+        5 / 0
+      }.check())
+      runner.report()
+    }.assert(r => r.failed == 1)
+
+    test("assert with 2 successful predicates") {
+      val runner = new Runner()
+      runner("test double") {
+        0.001
+      }.assert(_ >= 0.0, _ <= 1.0)
```
Comment (propensive) [inline src/test/tests.scala:151]:
```suggestion
      runner("test double")(0.001).assert(_ >= 0.0, _ <= 1.0)
```

File: src/test/tests.scala [hunk 18]
```diff
@@ -109,5 +111,62 @@ object Main extends Suite("Probably Tests") {
...
+        5 / 2
+      }.check()
+      runner.report()
+    }.assert(r => r.passed == r.total)
+
+    test("check without predicate should fail") {
+      val runner = new Runner()
+      Try(runner("test division") {
+        5 / 0
+      }.check())
```
Comment (propensive) [inline src/test/tests.scala:143]:
```suggestion
      Try(runner("test division")(5/0).check())
```

File: src/test/tests.scala [hunk 19]
```diff
@@ -109,5 +111,62 @@ object Main extends Suite("Probably Tests") {
...
+        5 / 0
+      }.assert()
+      runner.report()
+    }.assert(r => r.passed == 0)
+
+    test("check without predicate should succeed") {
+      val runner = new Runner()
+      runner("test division") {
+        5 / 2
+      }.check()
```
Comment (propensive) [inline src/test/tests.scala:135]:
```suggestion
      runner("test division")(5/2).check()
```

File: src/test/tests.scala [hunk 20]
```diff
@@ -109,5 +111,62 @@ object Main extends Suite("Probably Tests") {
...
+        5 / 2
+      }.assert()
+      runner.report()
+    }.assert(r => r.passed == r.total)
+
+    test("assert without predicate should fail") {
+      val runner = new Runner()
+      runner("test division") {
+        5 / 0
+      }.assert()
```
Comment (propensive) [inline src/test/tests.scala:127]:
```suggestion
      runner("test division")(5/0).assert()
```

File: src/test/tests.scala [hunk 21]
```diff
@@ -109,5 +111,62 @@ object Main extends Suite("Probably Tests") {
       }.assert(_ == 2)
       runner.report
     }.assert(_.results.head.count == 100)
+
+    test("assert without predicate should succeed") {
+      val runner = new Runner()
+      runner("test division") {
+        5 / 2
+      }.assert()
```
Comment (propensive) [inline src/test/tests.scala:119]:
```suggestion
      runner("test division")(5/2).assert()
```

**Review by propensive — CHANGES_REQUESTED** [review]
I think the tests could look better if we wrote a small helper method that gets the report from a new `Runner`, i.e.
```scala
def run(fn: Runner => Unit): Report = {
  val run = new Runner()
  fn(run)
  run.report()
}
```

we could tidy up quite a lot of common code in these tests.

File: src/core/probably.scala (hunk 5 above)
Comment (wi101) [reply src/core/probably.scala:139]: there is another instruction in the next line, should we leave it that way?
```scala
case Success(value) =>
          record(this, time, makeDatapoint(predicates, 0, Pass, value))
          value
```

**Review by propensive — APPROVED** [review]

**Pull request merged**